---
layout: subpage
---


<!DOCTYPE html>
<html>
	<body>

		<h1>Usage</h1>
		
<h2>1. Running an existing model </h2>
<h3 id="sec1.1">1.1 How to run an existing model?</h3>
<p>
Open a console at the root of the RBApy repository.
</p>
<p>
If the RBA model already exists,  run:
</p>
<pre><code> python solve_rba_model.py path/to/model </code></pre>

where the path points to the directory containing the XML files defining the RBA model.



<h3 id="sec1.2">1.2 How to visualize results of RBApy simulations?</h3>

<h3 id="sec1.3">1.3 How to change the medium?</h3>

<p>
The medium composition is defined in the file <b><font color="#77B5FE"> medium.tsv</font></b>. The file is located in the  directory containing the XML files defining the RBA model. 
The first column contains the prefixes of metabolites that are detected in the external compartments in the SBML (see section 5.1). 
The second column corresponds to the concentration of metabolites in millimoles.
</p>

<p>
To change the medium composition:
	<ul>
		<li>Open the file medium.tsv defining the medium composition. </li>
		<li>Update the metabolite concentration as follows:</li>
		
</p>

<pre><code> --------------------------------
Original file
--------------------------------
Metabolite    Concentration
M_k    10
M_no3    10
etc. </code></pre>

<p>
<li>Change the concentration of M_no3 from 10mM to 5mM and save the file.</li>
</p>
<pre><code> --------------------------------
New file
--------------------------------
Metabolite    Concentration
M_k    10
M_no3    5
etc. </code></pre>
<p>
<li>Run again the model using:</li>
</p>
<pre><code> python solve_rba_model.py path/to/model </code></pre>

		</ul>
<p>
Here RBApy evaluates the apparent catalytic rates of transporters using the new metabolite concentrations, before solving the optimization problem.
</p>



<h3 id="sec1.4">1.4 Can I change my model parameters? </h3>
<p>
RBA model parameters are defined in the file <b><font color="#77B5FE"> parameters.xml</font></b>. This file is located in the  directory containing the XML files of the RBA model. 
</p>
<p>
To update parameters
Open the file parameters.xml
Update the value of the parameter that you want to change. For instance:
</p>	
<pre><code> --------------------------------
Original file: 
the function ribosomeEfficiencyMM describing the efficiencies of ribosomes
max(Ymin, kmax*µ/(Km+µ)) 
--------------------------------
  &lt;function id="ribosomeEfficiencyMM" type="michaelisMenten">
         &lt;listOfParameters>
                &lt;parameter id="kmax" value="97200"/>
                &lt;parameter id="Km" value="0.5"/>
                &lt;parameter id="Y_MIN" value="32400"/>
         &lt;/listOfParameters>
  &lt;/function>
  etc. </code></pre>
<p>
Change the maximal rate of translation from 97200 aa/h to 48600 aa/h:
</p>
<pre><code> --------------------------------
New file: 
the function ribosomeEfficiencyMM describing the efficiencies of ribosomes
max(Ymin, kmax*µ/(Km+µ)) 
--------------------------------
  &lt;function id="ribosomeEfficiencyMM" type="michaelisMenten">
         &lt;listOfParameters>
                &lt;parameter id="kmax" value="48600"/>
                &lt;parameter id="Km" value="0.5"/>
                &lt;parameter id="Y_MIN" value="32400"/>
         &lt;/listOfParameters>
  &lt;/function>></code></pre>
<p>
Run the model using:</p>
<pre><code> python solve_rba_model.py path/to/model </code></pre>
<p>
Here RBApy will automatically read the new values. For the units of all parameters, please refer to the paper Goelzer et al. 2015, or to the description of the XML format defining RBA models.
</p>
<p>
 <b><font color="#77B5FE">Be careful when you update some parameters! </font></b>
Some parameters such as the fraction of total proteins allocated to the different compartments must sum to 1.
The fraction of non-enzymatic proteins within compartments should be lower than 1.
</p>
	
	
<h2>2. Creating a new RBA model </h2> 
    
<h3 id="sec2.1">2.1 What do I need to create a new RBA model?</h3>
To create your own RBA model, you need the full rbapy package, and:
<ul>
<li>The NCBI taxon ID of your organism (e.g. 83333 for Escherichia coli K-12) </li>
<li>a valid SBML file for the same organism (see requirements in section 5.1).</li>
</ul>
Thanks to the taxon ID, RBApy will automatically download the latest release of your organism on Uniprot and will try to match Uniprot gene names with the gene names of the SBML file.

<h3 id="sec2.2">2.2 How to create my new RBA model?</h3>
Define one input directory and put the SBML file, the fasta files in the input directory.

Fill in the file input/params.in, where you will specify the taxon ID, the path to the input and output directories, the name of the SBML file, and the external compartments of the SBML file

Open a console at the root of the RBApy repository and run:
<pre><code>python generate_rba_model.py path/to/params.in</code></pre>

This execution generates:
<ul>
<li>several XML files describing the RBA model. They are located in the output directory specified in params.in</li>
<li>several CSV Helper files (see section 5.3) containing default values for  ambiguous information (e.g. cofactor IDs which couldn’t automatically be linked to model metabolites). They are located in the input directory, with your SBML and fasta files.</li> 
</ul>
You can now refine your model (see <a href="#sec2.3">section 2.3</a>) or run your model (see <a href="#sec1.1">section 1.1</a>).

<h3 id="sec2.3">2.3 How to refine my RBA model?</h3>
After the first run of model generation, one has:
<ul>
<li>several XML files describing the RBA model. They are located in the output directory specified in params.in</li>
<li>several CSV Helper files (see section 5.3) containing default values for  ambiguous information (e.g. cofactor IDs which couldn’t automatically be linked to model metabolites). They are located in the input directory, with your SBML and fasta files. </li>
</ul>

To refine an RBA model, the user can:
After the first run of model generation, one has:
<ul>
<li>Edit manually the XML files (not recommended) </li>
<li>Edit manually the fasta files and the helper files, fill them in with information and re-run the model generation using: </li>
<pre><code> python generate_rba_model.py path/to/params.in</code></pre>
</ul>
We strongly recommend the user to start the edition of helper and fasta files in the following order:
<ol>
<li>Update  the helper file location_map.tsv (<a href="#sec2.3.2.2">section 2.3.2.2</a>)</li>
<li>Update  the helper file metabolites.tsv (<a href="#sec2.3.2.1">section 2.3.2.1</a>)</li>
<li>Update the fasta files with user-defined molecular machineries (<a href="#sec2.3.1">section 2.3.1</a>)</li>
<li>Update helper files specific to proteins: location.tsv (<a href="#sec2.3.2.3">section 2.3.2.3</a>), cofactors.tsv  (<a href="#sec2.3.2.4">section 2.3.2.4</a>) and subunits.tsv  (<a href="#sec2.3.2.5">section 2.3.2.5</a>) </li>
</ol>
During step 3, the new proteins detected within the fasta files will be automatically added in files location.tsv, cofactors.tsv and  subunits.tsv if missing information is detected. Therefore doing step 3 before step 4 enables to edit these files only one time.

The user can regenerate the model each time a helper file is modified using 
<pre><code> python generate_rba_model.py path/to/params.in</code></pre>
Or only at the end.
The user can also choose to solve the model each time  to see the impact of these modifications on the simulations. 
<pre><code> python solve_rba_model.py path/to/model</code></pre>

<h3 id="sec2.3.1">2.3.1 Insertion of user-defined molecular machineries</h3>

<p>
During the first run of RBApy, the package will use by default the ribosome composition, 
the average chaperone composition and the tRNAs of <i>Escherichia coli</i>.
To adapt the machineries to the organism of interest, simply substitute the proteins/rnas within the files 
while keeping strictly the format.
</p>
<p>
An example: the ribosome machinery.
The ribosome is usually composed of 3rRNAs (5S, 15S,  23S), around 50 ribosomal proteins, 
and the initiation, elongation and release factors.
The <b>ribosome.fasta</b> lists all the machinery composition in a pseudo-fasta format:
</p>
<pre><code>>rba|rrnO-16S|rrnO-16S|rna|1
TTTATCGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCCTAATACATGCA
AGTCGAGCGGACAGATGGGAGCTTGCTCCCTGATGTTAGCGGCGGACGGGTGAGTAACAC
GTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCTAATACCGGATGG
TTGTTTGAA …
>rba|rrnO-23S|rrnO-23S|rna|1
GGTTAAGTTAGAAAGGGCGCACGGTGGATGCCTTGGCACTAGGAGCCGATGAAGGACGGG
ACGAACACCGATATGCTTCGGGGAGCTGTAAGCAAGCTTTGATCCGGAGATTTCCGAATG
GGGAAACC …
>rba|rrnO-5S|rrnO-5S|rna|1
TTTGGTGGCGATAGCGAAGAGGTCACACCCGTTCCCATACCGAACACGGAAGTTAAGCTC
TTCAGCGCCGATGGTAGTCGGGGGTTTCCCCCTGTGAGAGTAGGACGCCGCC
>rba|P0AG67|RS1_ECOLI 30S ribosomal protein S1 OS=Escherichia coli (strain K12) GN=rpsA PE=1 SV=1|protein|1
MTESFAQLFEESLKEIETRPGSIVRGVVVAIDKDVVLVDAGLKSESAIPAEQFKNAQGEL
EIQVGDEVDVALDA…
…</code></pre>
The format of entry is as follows:
<pre><code>>rba | Molecule ID | Molecule Name | Molecule Type | Molecule Stoichiometry
--- Sequence ----
</code></pre>
<ul>
<li>rba: means that this entry corresponds to RBA machinery</li>
<li>Molecule ID: The identifier of the molecules. For proteins, the ID must corresponds either to the Uniprot Protein, either to one of the gene names.</li>  
<li>Molecule Name: the name of the molecule.  </li>
<li>Molecule Type: the Type of the molecule. Type can take two values: protein, rna. See below. The molecule type enables to insert automatically the new molecule 
in proteins.xml and rnas.xml, and also 
in processes.xml, within the list of production of processes that target molecules of same Type </li>
<li>Molecule Stoichiometry: the stoichiometry of the molecule within the molecular machine.</li>
</ul>

<p>
During the parsing of the fasta file, each time a protein is read, RBApy tries to extract automatically the location (see <a href="#sec2.3.2.3">section 2.3.2.3</a>), cofactors (see <a href="#sec2.3.2.4">2.3.2.4</a>) of the protein. In case of ambiguities, the protein is also inserted in the corresponding files locations.tsv, and cofactors.tsv. 
If the protein is not present in the uniprot file, the protein is still inserted in proteins.xml using the amino-acid sequence of the fasta file, and in processes.xml.  By default, the protein will be located in the cytoplasm, without cofactors. Moreover, the unknown protein is also inserted in the file unknown_proteins.tsv (see <a href="#sec2.3.2.6">2.3.2.6</a>).
</p>

<p>
An advice: For the ID of proteins, we recommend to use either the Uniprot protein ID (e.g. P21464), either the absolute identifier of the gene name (e.g. BSU16490), but not the gene name in 4 letters (e.g. rpsB). Indeed, the field gene name in Uniprot contains the current gene name of the protein, but also all synonyms. Sometimes, some genes may have been renamed, which can result in multiple entries for the 4-letters gene name.
</p>

To update the RBA model, re-run the model generation by
<pre><code>python generate_rba_model.py path/to/params.in</code></pre>

Now look in the files
<ul>
<li>proteins.xml which contains the new proteins of the molecular machine.</li>
<li>rnas.xml  which contains the new rnas of the molecular machine</li>
<li>processes.xml where the new proteins have been included in the processes “Translation” and “Folding”, and the new rnas in the processes “Transcription” and “RNA degradation”</li>
</ul>


<h3 id="sec2.3.2">2.3.2 Helper files</h3>
		       
<h4 id="sec2.3.2.1">2.3.2.1 Metabolites.tsv</h4>
<p>
The file <b><font color="#77B5FE">metabolites.tsv</font></b> is critical for RBApy. First It identifies metabolic precursors 
	that are consumed/produced by the cellular processes to build biomass component. 
	Second, the file is also used to link these metabolic biobricks to SBML ids of metabolites from the metabolic network. 
</p>	

In metabolites.tsv, the user declare:
<ol>
<li>M1: All metabolic precursors that are consumed/produced by the cellular processes to build proteins, RNAs, DNAs. A minima, RBApy expects to find the metabolic precursors involved in the default processes: Replication, Transcription, Translation, Folding, RNA degradation, which are: all charged/uncharged tRNAs, NMP, NDP, NTP, dNTP, water, protons, mono- and di-phosphate.
Therefore, RBApy tries to detect common SBML IDs for these precursors, namely:
<ul>
<li>M_trnaala_c, M_trnaarg_c, ..., M_trnaval_c for uncharged tRNAs</li>
<li>M_alatrna_c, M_argtrna_c, ..., M_valtrna_c for charged tRNAs</li>
<li>M_ATP_c, M_GTP_c, M_CTP_c, M_UTP_c, for ATP, GTP, CTP and UTP respectively</li>
<li>M_ADP_c, M_GDP_c, M_CDP_c, M_UDP_c, for ADP, GDP, CDP and UDP respectively</li>
<li>M_AMP_c, M_GMP_c, M_CMP_c, M_UMP_c, for AMP, GMP, CMP and UMP respectively</li>
<li>M_h2o_c, M_h_c, M_p_c, M_pp_c for water, proton, mono- and di-phosphate respectively</li>
	</ul></li>
<li>M2: All ions/cofactors required to have the active form of a protein. The ions and cofactors are automatically extracted from the Uniprot annotation (see <a href="#sec2.3.2.5">section 2.3.2.5</a>) and reported in metabolites.tsv for ID matching.</p>
 </li>
<li>M3: All metabolites for which the intracellular concentration is growth-rate independent, and the value of this concentration. The set of growth-rate independent metabolites includes building 
	blocks for cell wall and membrane synthesis, a few metabolites such as co-enzyme A. Usually these metabolites are present in the biomass reaction of the SBML file. Exclude all metabolites from the set M1 and M2 from the biomass reaction. The remaining ones form the set M3. The targeted concentrations correspond to the stoichiometric coefficients of the biomass reaction.
</li>
</ol>
By default, the file is built using the sets M1, M2 and M3 of the bacterium <i>Bacillus subtilis</i>.

<p>
The file metabolites.tsv, is organized in 4 columns, entitled ID, NAME, SBML ID and  CONCENTRATION.
<ul>
<li>ID: is the absolute identifier of the metabolite, which can be the ChEBI;</li>
<li>NAME: the name of the metabolite;</li>
<li>SBML ID: the SBML ID corresponding to the metabolite, present in the genome-scale metabolic model;</li>
<li>CONCENTRATION: the concentration of growth-rate independent metabolite. </li>
</ul>
To update the RBA model, re-run the model generation by
<pre><code>python generate_rba_model.py path/to/params.in</code></pre>
</p>
	
<p>
Now look in the files
<ul> 
<li>processes.xml: The lists of of metabolic precursors used in cellular processes are updated </li>
<li>targets.xml: The targeted group named metabolite_production is updated with the set of metabolite M3, and their targeted concentrations are reported in the file parameters.xml. </li>
	</ul>
</p>

<h4 id="sec2.3.2.2">2.3.2.2 Macrocomponents.tsv (optional) </h4>
<p>
	The file <b><font color="#77B5FE">macrocomponents.tsv</font></b> describes the metabolic requirements to build macro-components such as the cell wall or the cytoplasmic membrane. Usually, the intracellular concentration of such macro-components is growth-rate independent. The file describes which metabolite is required for macro-component synthesis, and  the value of this concentration. 
</p>
<p>
The file is organized in two columns, TARGET_METABOLITE  and  TARGET_FLUX.
<ul>
<li>TARGET_METABOLITE: the SBML ID of the metabolic precursor necessary to build the macro-component;</li>
<li> TARGET_FLUX: the value C of the targeted concentration.</li>
</ul>
The pipeline will automatically add to each metabolic precursor  a consumption flux equal to µC, where µ is the growth rate.
</p>
<p>
To update the RBA model, re-run the model generation by
<pre><code>python generate_rba_model.py path/to/params.in</code></pre>
Remark: This file is optional, because the identification of metabolic precursors and their targeted concentrations can also be defined in metabolites.tsv. However, the user may choose to define these precursors in a separate file, because of the specific role of macro-components in the cell.
</p>


<h4 id="sec2.3.2.3">2.3.2.3 Location_map.tsv </h4>
<p>
The file location_map.tsv is used to link uniprot location ids with user-defined compartment ids. For every uniprot location, user may fill in a name compartment (e.g. SBML compartment names). 
The idea behind this file is that it can be used to fuse compartments or ignore them. 
If the field is left empty (or [MISSING]), a compartment named after uniprot location is created.
</p>

An example:
<table>
<tr><td>UNIPROT NAME</td>    <td>USER ID</td></tr>
<tr><td>Cell outer membrane</td>    <td>Cell_outer_membrane</td></tr>
<tr><td>Cell inner membrane</td>    <td>Cell_inner_membrane</td></tr>
<tr><td>Cytoplasm</td>    <td>Cytoplasm</td></tr>
<tr><td>Cell membrane</td>   <td>Cell_inner_membrane</td></tr>
<tr><td>Periplasm</td>    <td>Periplasm</td></tr>
<tr><td>Membrane</td>    <td>Cell_inner_membrane</td></tr>
<tr><td>Secreted</td>    <td>Secreted</td></tr>
</table>
<p>
In this example, proteins that have the location "Cell membrane" and "Membrane" have the same location in the RBA model: "Cell_inner_membrane"
</p>
<p>
To update the RBA model, re-run the model generation by
</p>
<pre><code> python generate_rba_model.py path/to/params.in</code></pre>

Now look in the file metabolism.xml that contains the list of cellular compartments.  Compartments are updated as follows:
<pre><code> 
&lt;RBAMetabolism>
  &lt;listOfCompartments>
    &lt;compartment id="Cell_inner_membrane"/>
    &lt;compartment id="Cytoplasm"/>
    &lt;compartment id="Cell_outer_membrane"/>
    &lt;compartment id="Periplasm"/>
    &lt;compartment id="Secreted"/>
  &lt;/listOfCompartments>
</code></pre>

	</body>
</html>
